azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,340 @@
+            prefix_msg = msg_.split("incarnation", 1)[0]
+            prefix_last_warning_msg = self.persistent_data.last_warning.split("incarnation", 1)[0]
+            if prefix_msg != prefix_last_warning_msg or self.persistent_data.last_warning_time == datetime.datetime.min or now >= self.persistent_data.last_warning_time + datetime.timedelta(hours=6):
+                if level == LogLevel.WARNING:
+                    logger.warn(msg_)
+                elif level == LogLevel.ERROR:
+                    logger.error(msg_)
+                add_event(op=WALAEventOperation.AgentUpgrade, is_success=success_, message=msg_, log_event=False)
+                self.persistent_data.last_warning_time = now
+                self.persistent_data.last_warning = msg_
+
+    def run(self, goal_state):
+        try:
+            # Ignore new agents if update is disabled. The latter flag only used in e2e tests.
+            if not self._autoupdate_enabled or not conf.get_ga_updates_enabled():
+                return
+
+            self._gs_id = goal_state.extensions_goal_state.id
+            agent_family = self.__get_agent_family_from_last_gs(goal_state)
+            requested_version = self.__get_requested_version(agent_family)
+            agent_manifest = None  # This is to make sure fetch agent manifest once per update
+
+            if requested_version is None:
+                if conf.get_enable_ga_versioning():  # log the warning only when ga versioning is enabled
+                    warn_msg = "Missing requested version in agent family: {0} for incarnation: {1}, fallback to largest version update".format(self._ga_family, self._gs_id)
+                    self.__log_event(LogLevel.WARNING, warn_msg)
+                    GAUpdateReportState.report_error_msg = warn_msg
+                agent_manifest = goal_state.fetch_agent_manifest(agent_family.name, agent_family.uris)
+                requested_version = self.__get_largest_version(agent_manifest)
+                self._is_requested_version_update = False
+            else:
+                self._is_requested_version_update = True
+                # Save the requested version to report back
+                GAUpdateReportState.report_expected_version = requested_version
+                # Remove the missing requested version warning once requested version becomes available
+                if "Missing requested version" in GAUpdateReportState.report_error_msg:
+                    GAUpdateReportState.report_error_msg = ""
+
+            if requested_version == CURRENT_VERSION:
+                return
+
+            # Check if an update is allowed
+            if not self.__should_update_agent(requested_version):
+                return
+
+            msg_ = "Goal state {0} is requesting a new agent version {1}, will update the agent before processing the goal state.".format(
+                self._gs_id, str(requested_version))
+            self.__log_event(LogLevel.INFO, msg_)
```

Comment:
periodic messages have produced issues in the past. we should avoid them (i've been removing them as i touch code)

if the messages change as the code evolves, it easy to miss that, and it can create very large logs